docs/optimizing-javascript-code-loading.md

```diff
@@ -0,0 +1,222 @@
+---
+id: optimizing-javascript-code-loading
+title: Optimizing JavaScript code loading
+---
+
+import Tabs from '@theme/Tabs'; import TabItem from '@theme/TabItem'; import constants from '@site/core/TabsConstants';
+
+Parsing and running JavaScript code requires memory and time. Because of this, as your app grows, it's often useful to delay loading code until it's needed for the first time. React Native comes with some standard optimizations that are on by default, and there are techniques you can adopt in your own code to help React load your app more efficiently. There are also some advanced automatic optimizations (with their own tradeoffs) that are suitable for very large apps.
+
+## Recommended: Use Hermes
+
+Hermes is the default engine for new React Native apps, and is highly optimized for efficient code loading. In release builds, JavaScript code is fully compiled to bytecode ahead of time. Bytecode is loaded to memory on-demand and does not need to be parsed like plain JavaScript does.
+
+:::info
+Read more about using Hermes in React Native [here](./hermes).
+:::
+
+## Recommended: Lazy-load large components
+
+If a component with a lot of code/dependencies is not likely to be used when initially rendering your app, you can use React's [`lazy`](https://react.dev/reference/react/lazy) API to defer loading its code until it's rendered for the first time. Typically, you should consider lazy-loading "screen"-level components in your app, so that adding new screens to your app does not increase its startup time.
+
+:::info
+Read more about [lazy-loading components with Suspense
+](https://react.dev/reference/react/lazy#suspense-for-code-splitting), including code examples, in React's documentation.
+:::
+
+### Tip: Avoid module side effects
+
+Lazy-loading components can change the behavior of your app if your component modules (or their dependencies) have _side effects_, such as modifying global variables or subscribing to events outside of a component. Most modules in React apps should not have any side effects.
+
+```tsx title="SideEffects.tsx"
+import Logger from './utils/Logger';
+
+// Side effect! This must be executed before React can even begin to render
+// the SplashScreen component, and can unexpectedly break code elsewhere in
+// your app if you later decide to lazy-load SplashScreen.
+global.logger = new Logger();
+
+export function SplashScreen() {
+  // ...
+}
+```
+
+## Advanced: Call `require` inline
```

Comment:
Should we callout and link to the Pitfalls section below?

Reply:
I'm treating those as pitfalls of *automatic* inline requires. With manual inlining (which is already called out as "advanced"), you are in control, so if your code doesn't work, you can dig into why; you aren't dealing with surprising breakages "at a distance" like you would be with automatic inlining.